src/libraries/System.Private.CoreLib/src/System/ArgumentOutOfRangeException.cs

```diff
@@ -85,5 +88,85 @@ public override string Message
 
         // Gets the value of the argument that caused the exception.
         public virtual object? ActualValue => _actualValue;
+
+        [DoesNotReturn]
+        private static void Throw(string? paramName, string message)
+        {
+            throw new ArgumentOutOfRangeException(paramName, message);
+        }
+
+        /// <summary>Throws an <see cref="ArgumentOutOfRangeException"/> if <paramref name="value"/> is zero.</summary>
+        /// <param name="value">The argument to validate as non-zero.</param>
+        /// <param name="paramName">The name of the parameter with which <paramref name="value"/> corresponds.</param>
+        public static void ThrowIfZero<T>(T value, [CallerArgumentExpression(nameof(value))] string? paramName = null)
+            where T : INumberBase<T>
+        {
+            if (T.IsZero(value))
+                Throw(paramName, SR.Format(SR.ArgumentOutOfRange_Generic_MustBeNonZero, paramName));
```

Comment:
These ThrowXx methods need to be super streamlined, as we want them to be inlined and not contribute to bloat at the call sites.  As such, any use of SR, formatting, etc. needs to be pushed into the non-inlined helper that's doing the actual exception construction and throw.